src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestRunConfigurationProducerBase.kt

```diff
@@ -8,98 +8,144 @@ package org.rust.cargo.runconfig.test
 import com.intellij.execution.actions.ConfigurationContext
 import com.intellij.openapi.actionSystem.LangDataKeys
 import com.intellij.openapi.util.Ref
+import com.intellij.openapi.util.text.StringUtil
+import com.intellij.psi.PsiDirectory
 import com.intellij.psi.PsiElement
 import org.rust.cargo.project.workspace.CargoWorkspace
 import org.rust.cargo.runconfig.command.CargoCommandConfiguration
 import org.rust.cargo.runconfig.command.CargoRunConfigurationProducer
 import org.rust.cargo.runconfig.mergeWithDefault
 import org.rust.cargo.toolchain.CargoCommandLine
+import org.rust.lang.core.psi.RsFile
 import org.rust.lang.core.psi.RsFunction
 import org.rust.lang.core.psi.ext.*
 
-abstract class CargoTestRunConfigurationProducerBase : CargoRunConfigurationProducer() {
-    abstract val commandName: String
+private typealias TestConfigProvider = (List<PsiElement>, Boolean) -> TestConfig?
 
-    abstract fun isSuitable(element: RsElement): Boolean
+abstract class CargoTestRunConfigurationProducerBase : CargoRunConfigurationProducer() {
+    protected abstract val commandName: String
+    private val testConfigProviders: MutableList<TestConfigProvider> = mutableListOf()
+
+    init {
+        registerConfigProvider { elements, climbUp -> createConfigFor<RsFunction>(elements, climbUp) }
+        registerConfigProvider { elements, climbUp -> createConfigFor<RsMod>(elements, climbUp) }
+        registerConfigProvider { elements, climbUp -> createConfigForMultipleFiles(elements, climbUp) }
+        registerConfigProvider { elements, _ -> createConfigForDirectory(elements) }
```

Comment:
Maybe it's better to add more specific api for single directory?
Now we have three similar methods with the following code:
```kotlin
private fun createConfigFor*(elements: List<PsiElement>): TestConfig? {
    val dir = elements.singleOrNull() as? PsiDirectory ?: return null
    // some other code
}
```
We can create `registerDirectoryConfigProvider` method where we check that we have only one psi element is in context and it's a directory. It allows us to pass only single `PsiDirectory`  element into `createConfigFor*` methods, drop the same code in these methods and in general provide clearer API